### PR TITLE
Jwt authentication

### PR DESCRIPTION
**Pull request by Jawayria**
Added JWT authentication

### CODE REVIEW

File: cricPredict/Contest/views.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 class LeagueAPIView(APIView):
     permission_classes = [IsAuthenticated]
+    authentication_classes = [JWTAuthentication]
```
Comment (mraarif) [inline cricPredict/Contest/views.py]: I thought we agreed on configuring the authentication in a way that we don't need to specify `authentication_classes` on each view

File: cricPredict/Groups/views.py [hunk 2]
```diff
@@ -17,8 +17,8 @@ class GroupAPIView(APIView):
     queryset = ''
 
     def post(self, request):
-        print(self.request.user)
-        serializer = GroupSerializer(data=self.request.data, context={'user': self.request.user})
+        print(request.user)
```
Comment (mraarif) [inline cricPredict/Groups/views.py]: we should remove these and cleanup once we're done debugging

File: cricPredict/User_profile/TokenValidationMiddleware.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from django.core.cache import cache
+from django.core.exceptions import ValidationError
+
+
+class BaseMiddleware(object):
+    def __init__(self, get_response):
```
Comment (mraarif) [inline cricPredict/User_profile/TokenValidationMiddleware.py]: why do we need this empty class? can you instead explore `MiddlewareMixin`?

File: cricPredict/User_profile/urls.py [hunk 4]
```diff
@@ -1,12 +1,10 @@
...
-from . import views
+from .views import UserAPIView, Login, Logout, Signup
 
 urlpatterns = [
-    url(r'^user/$', UserAPIView.as_view(), name="User"),
-    url(r'^user/(?P<pk>[\d]+)/$', UserAPIView.as_view(), name="League"),
-    url(r'^$', Login.as_view(), name="Login"),
+    url(r'^list/$', UserAPIView.as_view(), name="User"),
+    url(r'^signup/$', Signup.as_view(), name='Signup'),
+    url(r'^methods/(?P<pk>[\d]+)/$', UserAPIView.as_view(), name="User APi"),
```
Comment (mraarif) [inline cricPredict/User_profile/urls.py]: why is this URL named methods? a URL should always represent some business object/use case

File: cricPredict/User_profile/urls.py [hunk 5]
```diff
@@ -1,12 +1,10 @@
...
+    url(r'^login/$', Login.as_view(), name="Login"),
```
Comment (mraarif) [inline cricPredict/User_profile/urls.py:8]: let's remove this in the next iteration, and move all API URLs under `api/`

File: cricPredict/cricPredict/settings.py [hunk 6]
```diff
@@ -40,12 +38,14 @@
     'django.contrib.messages',
     'django.contrib.staticfiles',
     'rest_framework',
+    'rest_framework.authtoken',
```
Comment (mraarif) [inline cricPredict/cricPredict/settings.py]: is this needed after we've `simple_jwt` integrated?

File: cricPredict/cricPredict/settings.py [hunk 7]
```diff
@@ -137,11 +133,24 @@
...
-            'rest_framework.authentication.BasicAuthentication',
-        )
-}
\ No newline at end of file
+    'DEFAULT_PERMISSION_CLASSES': [
+        'rest_framework.permissions.IsAuthenticated',
+        'rest_framework.permissions.AllowAny',
+    ],
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        "rest_framework.authentication.SessionAuthentication",
```
Comment (mraarif) [inline cricPredict/cricPredict/settings.py]: we should remove `SessionAuthentication` to disallow any old cookie-based login, you can verify this if you login into admin and then visit any API endpoint inside the browser, it'll show you DRF's browsable API and show you that the user is logged in(the one you logged in from admin).

File: cricPredict/cricPredict/urls.py [hunk 8]
```diff
@@ -16,11 +16,15 @@
...
 from django.contrib import admin
 from django.urls import path, include
+from rest_framework_simplejwt.views import TokenObtainPairView, TokenRefreshView
 
 urlpatterns = [
     path('admin/', admin.site.urls),
-    path('', include('User_profile.urls')),
-    path('group/', include('Groups.urls')),
-    path('contest/', include('Contest.urls')),
+    path('api/user/', include('User_profile.urls')),
```
Comment (mraarif) [inline cricPredict/cricPredict/urls.py:23]: oh, you've already done that, good

**Review by mraarif — COMMENTED** [review]
few suggestions/questions otherwise, looks good

File: cricPredict/Contest/views.py (hunk 1 above)
Comment (Jawayria) [reply cricPredict/Contest/views.py]: I forgot to remove these from views

File: cricPredict/User_profile/TokenValidationMiddleware.py (hunk 3 above)
Comment (Jawayria) [reply cricPredict/User_profile/TokenValidationMiddleware.py]: Okay

File: cricPredict/User_profile/urls.py (hunk 4 above)
Comment (Jawayria) [reply cricPredict/User_profile/urls.py]: This api has get, put, patch, delete methods. I couldn't think of some name which covers all these.

**Review by mraarif — APPROVED** [review]
Looks good enough, approving this, you might still want to read up on how to name/structure REST endpoint urls to match resources and Django's MiddlewareMixin

**Comment by mraarif** [comment]
Should be okay to merge

**Pull request merged**